pkg/apis/core/types.go

```diff
@@ -2765,6 +2769,20 @@ type PodDNSConfigOption struct {
 	Value *string
 }
 
+// An EphemeralContainer is a container which runs temporarily in a pod for human-initiated actions
+// such as troubleshooting. This is an alpha feature enabled by the EphemeralContainers feature flag.
+type EphemeralContainer struct {
+	// Spec describes the Ephemeral Container to be created.
+	Spec Container
```

Comment:
Perhaps embed this (just `Container`) so you don't add a layer of nesting.

Reply:
I like this suggestion. I think it improves the API and it solves the merge problem. It might not meet the requirement of having a v1.Container available for inspection by admission controllers. @liggitt wdyt?

Reply:
> It might not meet the requirement of having a v1.Container available for inspection by admission controllers. @liggitt wdyt?

I don't see an issue from that perspective. All the data is still there, in the expected structure. Any decoder that can't handle inlined types is already going to have a bad time with kube types (TypeMeta).

> I think it improves the API and it solves the merge problem.

@lavalamp @apelisse can the merge key logic handle a key that is one level deep in json but two levels deep in the go struct?

Reply:
> @lavalamp @apelisse can the merge key logic handle a key that is one level deep in json but two levels deep in the go struct?

Phrased this way I think it should work, but it doesn't necessarily sound like a good idea, and I'd like to look into more details to make sure it makes sense :-)

Phrased this way I think it should work: To elaborate on that: we use the openapi definition on a json object (not go type) to merge, so technically the merge code is unaware of the embedding etc.

Reply:
I gave it a quick try and it works, though I haven't tested it much. List order wasn't preserved in the way I expected, but I don't know the defined behavior.

@apelisse should I upload what I have so you can take a look?

Reply:
I can have a look

Reply:
I can't find an example in the core/v1 types of embedded list with attributes and I'm not completely convinced it's going to work. @verb can you maybe try to embed the `Container` and also use a reasonable `patchStrategy` and `patchMergeKey` on it, and see what openapi spec it generates?

Reply:
@apelisse I just pushed https://github.com/kubernetes/kubernetes/pull/59416/commits/c54eafbcdcc4d503baa0cb4eb0c97c482998f305 switching to an embedded container. Looks like tests pass.

[swagger.json](https://github.com/kubernetes/kubernetes/pull/59416/commits/1a53963fd6c6beada7644e2c494514ca3dffccd5#diff-9ce7ea8441086bf1902b4f936f4601d0) contains the fields of `Container` as if they were in `EphemeralContainer`.

What do you think?

Reply:
I think this should work, thanks!